File: bluefog/torch/handle_manager.cc
#include "handle_manager.h"

namespace bluefog {
namespace torch {

int HandleManager::AllocateHandle() {
  int handle = last_handle_.fetch_add(1) + 1;
  std::lock_guard<std::mutex> guard(mutex_);
  results_[handle] = nullptr;
  return handle;
}

void HandleManager::MarkDone(int handle, const Status& status) {
  std::lock_guard<std::mutex> guard(mutex_);
  results_[handle] = std::make_shared<Status>(status);
}

bool HandleManager::PollHandle(int handle) {
  std::lock_guard<std::mutex> guard(mutex_);
  if (results_.find(handle) == results_.end()) {
    throw std::invalid_argument("Handle " + std::to_string(handle) +
                                " was not created or has been cleared.");
  }
  return results_[handle] != nullptr;
}

std::shared_ptr<Status> HandleManager::ReleaseHandle(int handle) {
  std::lock_guard<std::mutex> guard(mutex_);
  if (results_.find(handle) == results_.end()) {
    throw std::invalid_argument("Handle " + std::to_string(handle) +
                                " was not created or has been cleared.");
  }
  auto status = results_[handle];
  results_.erase(handle);
  return status;
}

}  // namespace torch
}  // namespace bluefog

File: bluefog/torch/handle_manager.py
import  threading
from typing import Dict

from bluefog.common.common import Status


class HandleManager:
    _last_handle: int = -1
    _results: Dict[int, Status] = {}
    _lock: threading.Lock = threading.Lock()

    @classmethod
    def AllocateHandle(cls) -> int:
        with cls._lock:
            cls._last_handle += 1
            cls._results[cls._last_handle] = Status.INPROGRESS
        return cls._last_handle

    @classmethod
    def MarkDone(cls, handle) -> None:
        with cls._lock:
            cls._results[handle] = Status.OK

    @classmethod
    def PollHandle(cls, handle) -> bool:
        with cls._lock:
            if handle not in cls._results:
                raise ValueError(f"Handle {handle} was not created or has been cleared.")
            res = cls._results[handle] == Status.OK
        return res

    @classmethod
    def ReleaseHandle(cls, handle) -> Status:
        with cls._lock:
            if handle not in cls._results:
                raise ValueError(f"Handle {handle} was not created or has been cleared.")
            status = cls._results.pop(handle)
        return status
